src/v2/token.rs

```diff
@@ -0,0 +1,214 @@
+use std::{collections::HashMap, sync::Arc, time::SystemTime};
+
+use cache_control::CacheControl;
+use josekit::{jwk::JwkSet, jws::RS256, jwt, jwt::JwtPayload};
+use reqwest::{header, Client, Response};
+use serde::Deserialize;
+use time::OffsetDateTime;
+use tokio::sync::RwLock;
+use url::Url;
+
+/// Zitadel client to verify a token's validity
+#[derive(Debug)]
+pub struct ZitadelJWTVerifier {
+	/// Zitadel domain
+	domain: Url,
+	/// Client for performing the requests
+	client: Client,
+	/// Key set from Zitadel
+	jwks: Arc<RwLock<JwkSet>>,
+	/// Time when the jwks is no longer valid
+	expires_at: OffsetDateTime,
+}
+
+impl ZitadelJWTVerifier {
+	/// Creates a new verifier to verify with a specif server
+	#[must_use]
+	pub fn new(url: Url) -> Self {
```

Comment:
I thought about the idea of using the issuer on the token as the domain to validate the token. The problem with that is that the token can be valid on the issuer server and still don't come from our server. Any thoughts on that?

Reply:
Yes, you're right. We should just pass in an issuer and validate against that.

Reply:
I'm having problem to get url before I verify the token. I'm trying to take it from the iss claim, check it against a list of trusted servers and with that create the verifier. The jwt libs don't easily allow me to get the payload without verifying it. That makes me question if this is really the right approach. Maybe we should just verify the token with the server from the iss claim and the user of the ZitadelJWTVerifier should evaluate if the issuer is trusted or not. It seems a bit less secure to me because we delegate one of the checks to the user of the lib but it may make sense. What do you think @emgrav, @sirewix ?